examples/hdf5.py

```diff
@@ -4,99 +4,184 @@
+        return ds1 == ds2
+
+    def __getitem__(self, item):
+        if not isinstance(item, slice) or item.step is not None:
+            raise TypeError
+
+        i1, i2, ds, n = self._last = self._range(item.start, item.stop)
+
+        if ds < 4:
+            # downsampling has no effect
+            y = self.y[i1:i2]
+            i = np.arange(i1, i2)
         else:
-            # Here convert data into a down-sampled array suitable for visualizing.
-            # Must do this piecewise to limit memory usage.        
-            samples = 1 + ((stop-start) // ds)
-            visible = np.zeros(samples*2, dtype=self.hdf5.dtype)
-            sourcePtr = start
-            targetPtr = 0
-            
-            # read data in chunks of ~1M samples
-            chunkSize = (1000000//ds) * ds
-            while sourcePtr < stop-1: 
-                chunk = self.hdf5[sourcePtr:min(stop,sourcePtr+chunkSize)]
-                sourcePtr += len(chunk)
-                
-                # reshape chunk to be integral multiple of ds
-                chunk = chunk[:(len(chunk)//ds) * ds].reshape(len(chunk)//ds, ds)
-                
-                # compute max and min
-                chunkMax = chunk.max(axis=1)
-                chunkMin = chunk.min(axis=1)
-                
-                # interleave min and max into plot data to preserve envelope shape
-                visible[targetPtr:targetPtr+chunk.shape[0]*2:2] = chunkMin
-                visible[1+targetPtr:1+targetPtr+chunk.shape[0]*2:2] = chunkMax
-                targetPtr += chunk.shape[0]*2
-            
-            visible = visible[:targetPtr]
-            scale = ds * 0.5
-            
-        self.setData(visible) # update the plot
-        self.setPos(start, 0) # shift to match starting index
-        self.resetTransform()
-        self.scale(scale, 1)  # scale to match downsampling
+            if n < self._minCacheLevel:
+                y = downsample(self.y, i1, i2, ds)
+            else:
+                if hasattr(self, '_t'):
+                    self._t.join()
+                    del self._t
+                y = self._loadLevel(n)[2 * int(i1 / ds): 2 * int(i2 / ds) + 1]
+            i = np.arange(i1, i1 + len(y) * ds / 2, ds / 2)
+
+        x = i if self.x is None else self.x[i]
+        return x, y
+
+    def __len__(self):
+        return len(self.y)
+
+    def _range(self, x1, x2, pad=True):
+        if pad:
+            pad = (x2 - x1) * self.padding
+            x1, x2 = x1 - pad, x2 + pad
+        if self.x is None:
+            i1 = max(0, min(len(self), int(x1)))
+            i2 = max(0, min(len(self), int(x2)))
+        else:
+            i1, i2 = np.searchsorted(self.x, [x1, x2])
+        ds = max(1, int((i2 - i1) / self.sampleLimit))
+        n = int(math.log2(ds))
+        ds = 2 ** n
+        i1 = int(i1 / ds - 1) * ds
+        i2 = int(i2 / ds + 1) * ds
+        i1 = max(0, min(len(self), int(i1)))
```

Comment:
no need to cast `i1` to `int()` it should be an int from the previous operations